sklearn/base.py

```diff
@@ -319,7 +319,15 @@ def __getstate__(self):
             state = self.__dict__.copy()
 
         if type(self).__module__.startswith("sklearn."):
-            return dict(state.items(), _sklearn_version=__version__)
+            if "_sklearn_pickle_version" in state.keys():
+                pickle_version = state["_sklearn_pickle_version"]
+            else:
+                pickle_version = __version__
+            return dict(
+                state.items(),
+                _sklearn_version=__version__,
+                _sklearn_pickle_version=pickle_version,
+            )
```

Comment:
I do not think we need to add  another attribute. We can slightly adjust `__setstate__` to allow `_sklearn_version` to be part of the state:

```python
    def __setstate__(self, state):
        if type(self).__module__.startswith("sklearn."):
            try:
                pickle_version = state["_sklearn_version"]
            except KeyError:
                pickle_version = "pre-0.18"
        ...
```

This way the pickled version is an attribute after unpickling:

```python
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_iris
import pickle

iris = load_iris()
clf = DecisionTreeClassifier().fit(iris.data, iris.target)

pickled_clf = pickle.loads(pickle.dumps(clf))

# Version used to fitt estimator
pickled_clf._sklearn_version
```

A user can inspect then inspect `_sklearn_version` for the version the estimator was used to fitted on. If we go this route, I would prefer to rename `_sklearn_version` to `_sklearn_pickle_version` to be very explicity